agent/local/state.go

```diff
@@ -0,0 +1,1031 @@
+	// Therefore, we mark the service as deleted and keep the
+	// entry around until it is actually removed.
+	s.InSync = false
+	s.Deleted = true
+	l.TriggerSyncChanges()
+
+	return nil
+}
+
+// Service returns the locally registered service that the
+// agent is aware of and are being kept in sync with the server
+func (l *State) Service(id string) *structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return nil
+	}
+	return s.Service
+}
+
+// Services returns the locally registered services that the
+// agent is aware of and are being kept in sync with the server
+func (l *State) Services() map[string]*structs.NodeService {
+	l.RLock()
+	defer l.RUnlock()
+
+	m := make(map[string]*structs.NodeService)
+	for id, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Service
+	}
+	return m
+}
+
+// ServiceState returns a shallow copy of the current service state
+// record. The service record still points to the original service
+// record and must not be modified.
+func (l *State) ServiceState(id string) *ServiceState {
+	l.RLock()
+	defer l.RUnlock()
+
+	s := l.services[id]
+	if s == nil || s.Deleted {
+		return nil
+	}
+	return s.Clone()
+}
+
+// SetServiceState is used to overwrite a raw service state with the given
+// state. This method is safe to be called concurrently but should only be used
+// during testing. You should most likely call AddService instead.
+func (l *State) SetServiceState(s *ServiceState) {
+	l.Lock()
+	defer l.Unlock()
+
+	l.services[s.Service.ID] = s
+	l.TriggerSyncChanges()
+}
+
+// ServiceStates returns a shallow copy of all service state records.
+// The service record still points to the original service record and
+// must not be modified.
+func (l *State) ServiceStates() map[string]*ServiceState {
+	l.RLock()
+	defer l.RUnlock()
+
+	m := make(map[string]*ServiceState)
+	for id, s := range l.services {
+		if s.Deleted {
+			continue
+		}
+		m[id] = s.Clone()
+	}
+	return m
+}
+
+// CheckToken is used to return the configured health check token for a
+// Check, or if none is configured, the default agent ACL token.
+func (l *State) CheckToken(checkID types.CheckID) string {
+	l.RLock()
+	defer l.RUnlock()
+	return l.checkToken(checkID)
+}
+
+// checkToken returns an ACL token associated with a check.
+// This method is not synchronized and the lock must already be held.
+func (l *State) checkToken(id types.CheckID) string {
+	var token string
+	c := l.checks[id]
+	if c != nil {
+		token = c.Token
+	}
+	if token == "" {
+		token = l.tokens.UserToken()
+	}
+	return token
+}
+
+// AddCheck is used to add a health check to the local state.
+// This entry is persistent and the agent will make a best effort to
+// ensure it is registered
+func (l *State) AddCheck(check *structs.HealthCheck, token string) error {
+	if check == nil {
+		return fmt.Errorf("no check")
+	}
+
+	if l.discardCheckOutput.Load().(bool) {
+		check.Output = ""
```

Comment:
This feels like an unexpected side effect since you are passing in a pointer.

Reply:
If we wouldn't do this here then the first check might be stored with output but subsequent checks have their output discarded. In this case you'd update the server once even when the status hasn't changed and `discard_check_output=true`. Granted, this is in all likelihood only an issue during testing since only there you'll register a check with some initial output but that's how I stumbled over this.

I agree that this is a side effect but it is in one place (`AddCheck` and `UpdateCheck`). Otherwise, I'd have to move the discarding to the check registration in the API and the check execution.

If we do not want to modify the check record at all then we need to modify the compare and sync logic to push a clone of the record to the server which does not contain the output. I'm not sure that's worth it but I'll think about it.

Reply:
fixed.